### ISSUE TITLE
BOINC switching GPUs between tasks on multi-GPU systems

### ISSUE THREAD

**Issue by Azmodeszer** [BOINC/boinc#4670]
**Describe the bug**
When you are using more than one GPU for BOINC on a single system, the client tends to pointlessly switch cards between tasks after restarting (or perhaps it's random and there is a 50/50 etc. chance of the "wrong" card being used, I haven't looked into that exactly). This cannot be prevented in any way that I was able to figure out (you could run a client for each card or do it manually by suspending tasks and resuming them in such an order that it forces the tasks on the GPUs of your choosing, but that level of micromanagement is obviously not a desirable solution). The "Switch between tasks every X minutes" option has no effect on this, by the way. I have it set to the maximum of 9999 minutes (why is there no option to never switch anyway?), and while I guess it prevents switching during sessions for that theoretical amount of time, if you restart the client it is being ignored.

For some projects this isn't a problem, but GPUGrid for instance is very sensitive to switching GPUs. You cannot continue tasks on a different card. It will yield an immediate computation error in that case, often resulting in quite a lot of lost work (tasks can take up to a day or more, depending on the card).

**Steps To Reproduce**
1.  Attach to GPUGrid on a system with more than one GPU, crunch GPU tasks. Any GPU project will do, really, but you'll notice it right away with that one.
2.  Restart the client.
3. If the tasks error out you'll know that there's been a switch.

**Expected behavior**
Keep the same card on the same task from start to finish. There should be no need whatsoever to switch.

**Comment by davidpanderson** [comment]
BOINC assumes that all instances of a GPU type (e.g. VIDIA) are equivalent;
in particular, if a job starts on one GPU and writes a checkpoint file,
it can be resumed on a different GPU of the same type, using that checkpoint file.
The GPUGrid app apparently violates this assumption.
The first thing to do is to ask the GPUGrid developers why it does this.

**Comment by KeithMyers** [comment]
ERROR: /home/user/conda/conda-bld/acemd3_1566914012210/work/src/mdsim/context.cpp line 324: Cannot use a restart file on a different device!

So this has a failure in common with the Windows wrapper app. The BOINC wrapper app is the common element.

**Comment by RichardHaselgrove** [comment]
No, this error is specific to the current ACEMD3 application. I can't find your example task in your computer list, but the error message has been reported in context on the project message boards, most recently at https://www.gpugrid.net/forum_thread.php?id=5246&nowrap=true#58078
```
16:51:07 (662618): wrapper: running bin/acemd3 (--boinc --device 0)
ERROR: /home/user/conda/conda-bld/acemd3_1632842613607/work/src/mdsim/context.cpp line 318: Cannot use a restart file on a different device!
16:51:11 (662618): bin/acemd3 exited; CPU time 4.131044
```
The error causes acemd3 to exit, not the wrapper.

**Comment by KeithMyers** [comment]
Run the task with just the acemd3 app in a terminal.  Without the wrapper.
Suspend the task.  Close BOINC.  Wait 3 minutes.  Start BOINC back up and the task can resume without failure on a different class of device.
Leave the wrapper app out of the equation and all is well. Whatever communication happens between the wrapper app and the science app throws a monkey wrench into the works.

**Comment by RichardHaselgrove** [comment]
Interesting. Looking at the acemd3 application currently running on this machine (Windows 7/64, using Process Explorer), I see that the wrapper has launched acemd3 with the command line

`"bin\acemd3.exe" --boinc --device 0`

The wrapper translates the device specification issued by the boinc client in init_data.xml, into the command line format required by external applications that haven't been linked with the full BOINC communications library at compile time.

So maybe the problem is triggered by the `--boinc` parameter? I don't know what that does, in the acemd3 environment - we would have to ask the developers, or explore the acemd3 source code ourselves. The full repository isn't visible on GitHub, but there are clues like

`https://github.com/Acellera/swan/blob/d7e7daf8bd4ab3494a569f3bb98664de156002b8/src/deviceQuery.c#L103`

Could you explore in your terminal session?

**Comment by KeithMyers** [comment]
Just need to snag another on a dissimilar card host again.
